File: HLeaker - c#/HandleLeaker/Service/Service.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace HandleLeaker
{
    class Service
    {

        #region Structs
        public struct HANDLE_INFO
        {
            public int Pid;
            public IntPtr hProcess;
            public HANDLE_INFO(int Pid, IntPtr hProcess)
            {
                this.Pid = Pid;
                this.hProcess = hProcess;
            }
        }

        private struct HANDLE_IN
        {
            public IntPtr hObject;
            public bool PStatus;
            public bool IStatus;
            public IntPtr Function;
            public HANDLE_IN(IntPtr hObject, bool PStatus, bool IStatus, IntPtr Function)
            {
                this.hObject = hObject;
                this.PStatus = PStatus;
                this.IStatus = IStatus;
                this.Function = Function;
            }
        };

        private struct THREAD_IN
        {
            public IntPtr hProcess;
            public IntPtr _GetProcessId, _ExitThread;
            public THREAD_IN(IntPtr hProcess, IntPtr _GetProcessId, IntPtr _ExitThread)
            {
                this.hProcess = hProcess;
                this._GetProcessId = _GetProcessId;
                this._ExitThread = _ExitThread;
            }
        };

        #endregion Structs

        #region Members
        static IntPtr _ExitThread = IntPtr.Zero, _GetProcessId = IntPtr.Zero, _NtSetInformationObject = IntPtr.Zero;
        #endregion

        /// <summary>
        /// This function starts our target process in the normal session
        /// </summary>
        public static IntPtr ServiceStartProcess(string lpFile, string lpArguments, string lpDir, Boolean Inherit, IntPtr hParent)
        {
            Kernel32.STARTUPINFOEX si = new Kernel32.STARTUPINFOEX();
            Kernel32.PROCESS_INFORMATION pi = new Kernel32.PROCESS_INFORMATION();
            IntPtr processToken = IntPtr.Zero, userToken = IntPtr.Zero, pEnvironment = IntPtr.Zero, cbAttributeListSize = IntPtr.Zero, pAttributeList = IntPtr.Zero;
            Advapi32.SECURITY_ATTRIBUTES sa = new Advapi32.SECURITY_ATTRIBUTES();
            pi.hProcess = IntPtr.Zero;
            if (!Kernel32.OpenProcessToken(Kernel32.GetCurrentProcess(), Kernel32.TOKEN_ALL_ACCESS, ref processToken))
                goto EXIT;
            if (!Advapi32.DuplicateTokenEx(processToken, Kernel32.TOKEN_ALL_ACCESS, out sa, Advapi32.SECURITY_IMPERSONATION_LEVEL.SecurityImpersonation, Advapi32.TOKEN_TYPE.TokenPrimary, out userToken) ||
                !UserEnv.CreateEnvironmentBlock(ref pEnvironment, userToken, false))
                goto EXIT;
            Kernel32.InitializeProcThreadAttributeList(IntPtr.Zero, 1, 0, ref cbAttributeListSize);
            pAttributeList = Kernel32.VirtualAlloc(IntPtr.Zero, (int)cbAttributeListSize, 0x1000, 0x40);
            if (pAttributeList == null)
                goto EXIT;
            if (!Kernel32.InitializeProcThreadAttributeList(pAttributeList, 1, 0, ref cbAttributeListSize))
                goto EXIT;
            if (!Kernel32.UpdateProcThreadAttribute(pAttributeList, 0, (IntPtr)0x00020000, ref hParent, (IntPtr)Marshal.SizeOf(hParent), IntPtr.Zero, IntPtr.Zero))
                goto EXIT;
            si.lpAttributeList = pAttributeList;
            si.StartupInfo = new Kernel32.STARTUPINFO();
            si.StartupInfo.cb = Marshal.SizeOf(typeof(Kernel32.STARTUPINFO));
            if (!Kernel32.CreateProcessAsUserA(userToken, lpFile, lpArguments, IntPtr.Zero, IntPtr.Zero, Inherit, 0x400 | 0x010 |  0x00080000, pEnvironment, lpDir, ref si, ref pi))
                goto EXIT;
            EXIT:
            if(processToken != IntPtr.Zero)
                Kernel32.CloseHandle(processToken);
            if (userToken != IntPtr.Zero)
                Kernel32.CloseHandle(userToken);
            if (pEnvironment != IntPtr.Zero)
                UserEnv.DestroyEnvironmentBlock(pEnvironment);
            if (pAttributeList != IntPtr.Zero)
            {
                Kernel32.DeleteProcThreadAttributeList(pAttributeList);
                Kernel32.VirtualFree(pAttributeList, (int)cbAttributeListSize, 0x4000);
            }
            if(pi.hThread != IntPtr.Zero)
            Kernel32.CloseHandle(pi.hThread);
            return pi.hProcess;
        }

        /// <summary>
        /// This function executes a shellcode to change the handle characteristics
        /// </summary>
        public static Boolean ServiceSetHandleStatus(CProcess Process, IntPtr hObject, bool Protect, bool Inherit)
        {
            bool Is64 = false, Status = false;
            byte[] W64Thread = { 0x48, 0x83, 0xEC, 0x28, 0xF, 0xB6, 0x41, 0x8, 0x4C, 0x8D, 0x44, 0x24, 0x30, 0x41, 0xB9, 0x2, 0x0, 0x0, 0x0, 0x88, 0x44, 0x24, 0x31, 0xF, 0xB6, 0x41, 0xC, 0x4C, 0x8B, 0xD1, 0x48, 0x8B, 0x9, 0x88, 0x44, 0x24, 0x30, 0x41, 0x8D, 0x51, 0x2, 0x41, 0xFF, 0x52, 0x10, 0x33, 0xC9, 0x85, 0xC0, 0xF, 0x94, 0xC1, 0x8B, 0xC1, 0x48, 0x83, 0xC4, 0x28, 0xC3 };
            byte[] W32Thread = { 0x55, 0x8B, 0xEC, 0x8B, 0x4D, 0x8, 0x6A, 0x2, 0xF, 0xB6, 0x41, 0x4, 0x88, 0x45, 0x9, 0xF, 0xB6, 0x41, 0x8, 0x88, 0x45, 0x8, 0x8D, 0x45, 0x8, 0x50, 0x8B, 0x41, 0xC, 0x6A, 0x4, 0xFF, 0x31, 0xFF, 0xD0, 0xF7, 0xD8, 0x1B, 0xC0, 0x40, 0x5D, 0xC2, 0x4, 0x0 };
            HANDLE_IN Args;
            IntPtr hThread = IntPtr.Zero, lpArgs = IntPtr.Zero, lpThread = IntPtr.Zero, WThread = IntPtr.Zero, WArgs = IntPtr.Zero;
            if (hObject == IntPtr.Zero || Process.Is64(ref Is64) != 1)
                goto EXIT;
            if ((IntPtr.Size == 8) ? !Is64 : Is64)
                goto EXIT;
            if (_NtSetInformationObject == IntPtr.Zero)
            {
                _NtSetInformationObject = Kernel32.GetProcAddress(Kernel32.GetModuleHandleA("ntdll.dll"), "NtSetInformationObject");
                if (_NtSetInformationObject == IntPtr.Zero)
                    goto EXIT;
            }
            Args = new HANDLE_IN(hObject, Protect, Inherit, _NtSetInformationObject);
            if (Args.Function == IntPtr.Zero)
                goto EXIT;
            if ((lpThread = Kernel32.VirtualAllocEx(Process.GetHandle(), IntPtr.Zero, Is64 ? W64Thread.Length : W32Thread.Length, 0x1000, 0x40)) == IntPtr.Zero ||
                (lpArgs = Kernel32.VirtualAllocEx(Process.GetHandle(), IntPtr.Zero, Marshal.SizeOf(typeof(HANDLE_IN)), 0x1000, 0x40)) == IntPtr.Zero)
                goto EXIT;
            WArgs = Marshal.AllocHGlobal(Marshal.SizeOf(WArgs));
            WThread = Marshal.AllocHGlobal(Is64 ? W64Thread.Length : W32Thread.Length);
            Marshal.Copy(Is64 ? W64Thread : W32Thread, 0, WThread, Is64 ? W64Thread.Length : W32Thread.Length);
            Marshal.StructureToPtr(Args, WArgs, true);
            if (!Kernel32.WriteProcessMemory(Process.GetHandle(), lpThread, WThread, Is64 ? W64Thread.Length : W32Thread.Length, IntPtr.Zero) ||
                !Kernel32.WriteProcessMemory(Process.GetHandle(), lpArgs, WArgs, Marshal.SizeOf(Args), IntPtr.Zero))
                goto EXIT;

            if (ntdll.RtlCreateUserThread(Process.GetHandle(), IntPtr.Zero, false, IntPtr.Zero, IntPtr.Zero, IntPtr.Zero, lpThread, lpArgs, ref hThread, IntPtr.Zero) != 0)
                goto EXIT;
            Kernel32.WaitForSingleObject(hThread, (0xFFFFFFFF));
            Status = true;
            EXIT:
            if ((hThread) != IntPtr.Zero)
                Kernel32.CloseHandle(hThread);
            if (lpThread != null)
                Kernel32.VirtualFreeEx(Process.GetHandle(), lpThread, Is64 ? W64Thread.Length : W32Thread.Length, 0x4000);
            if (lpArgs != null)
                Kernel32.VirtualFreeEx(Process.GetHandle(), lpArgs, Marshal.SizeOf(typeof(HANDLE_IN)), 0x4000);
            return Status;
        }
        /// <summary>
        /// This function executes a shellcode to get the process id of an existing handle
        /// </summary>
        private static Boolean ServiceGetProcessId(CProcess Process, IntPtr Handle, ref uint ProcessId)
        {
            byte[] W32Thread = { 0x55, 0x8B, 0xEC, 0x56, 0x8B, 0x75, 0x08, 0xFF, 0x36, 0x8B, 0x46, 0x04, 0xFF, 0xD0, 0x50, 0x8B, 0x46, 0x08, 0xFF, 0xD0, 0x5E, 0x5D, 0xC3 };
            byte[] W64Thread = { 0x40, 0x53, 0x48, 0x83, 0xEC, 0x20, 0x48, 0x8B, 0xD9, 0x48, 0x8B, 0x09, 0xFF, 0x53, 0x08, 0x8B, 0xC8, 0xFF, 0x53, 0x10, 0xB8, 0x01, 0x00, 0x00, 0x00, 0x48, 0x83, 0xC4, 0x20, 0x5B, 0xC3 };
            bool Is64 = false, Status = false;
            IntPtr hThread = IntPtr.Zero, lpArgs = IntPtr.Zero, lpThread = IntPtr.Zero, WThread = IntPtr.Zero, WArgs = IntPtr.Zero, k32 = IntPtr.Zero;
            THREAD_IN Args;

            if (Handle == IntPtr.Zero || Process.Is64(ref Is64) != 1)
                goto EXIT;
            if ((IntPtr.Size == 8) ? !Is64 : Is64)
                goto EXIT;
            if (_ExitThread == IntPtr.Zero && _GetProcessId == IntPtr.Zero)
            {
                k32 = Kernel32.GetModuleHandleA("kernel32.dll");
                _ExitThread = Kernel32.GetProcAddress(k32, "ExitThread");
                if (_ExitThread == IntPtr.Zero)
                    goto EXIT;

                _GetProcessId = Kernel32.GetProcAddress(k32, "GetProcessId");
                if (_GetProcessId == IntPtr.Zero)
                    goto EXIT;
            }
            Args = new THREAD_IN(Handle, _GetProcessId, _ExitThread);
            if ((lpThread = Kernel32.VirtualAllocEx(Process.GetHandle(), IntPtr.Zero, Is64 ? W64Thread.Length : W32Thread.Length, 0x1000, 0x40)) == IntPtr.Zero ||
                (lpArgs = Kernel32.VirtualAllocEx(Process.GetHandle(), IntPtr.Zero, Marshal.SizeOf(typeof(THREAD_IN)), 0x1000, 0x40)) == IntPtr.Zero)
                goto EXIT;
            WArgs = Marshal.AllocHGlobal(Marshal.SizeOf(Args));
            WThread = Marshal.AllocHGlobal(Is64 ? W64Thread.Length : W32Thread.Length);
            Marshal.Copy(Is64 ? W64Thread : W32Thread, 0, WThread, Is64 ? W64Thread.Length : W32Thread.Length);
            Marshal.StructureToPtr(Args, WArgs, true);
            if (!Kernel32.WriteProcessMemory(Process.GetHandle(), lpThread, WThread, Is64 ? W64Thread.Length : W32Thread.Length, IntPtr.Zero) ||
                !Kernel32.WriteProcessMemory(Process.GetHandle(), lpArgs, WArgs, Marshal.SizeOf(Args), IntPtr.Zero))
                goto EXIT;
            if (ntdll.RtlCreateUserThread(Process.GetHandle(), IntPtr.Zero, false, IntPtr.Zero, IntPtr.Zero, IntPtr.Zero, lpThread, lpArgs, ref hThread, IntPtr.Zero) != 0)
                goto EXIT;
            if (Kernel32.WaitForSingleObject(hThread, Options.ObjectTimeout) == 0x00000102)
            {
                Kernel32.TerminateThread(hThread, 0x102);
                goto EXIT;
            }
            Status = Kernel32.GetExitCodeThread(hThread, out ProcessId);
            EXIT:
            if ((hThread) != IntPtr.Zero)
                Kernel32.CloseHandle(hThread);
            if (lpThread != null)
                Kernel32.VirtualFreeEx(Process.GetHandle(), lpThread, Is64 ? W64Thread.Length : W32Thread.Length, 0x4000);
            if (lpArgs != null)
                Kernel32.VirtualFreeEx(Process.GetHandle(), lpArgs, Marshal.SizeOf(typeof(THREAD_IN)), 0x4000);
            return Status;
        }
        /// <summary>
        /// This function enumrates the handles to our target process
        /// </summary>
        public static List<HANDLE_INFO> ServiceEnumHandles(int ProcessId, UInt32 DesiredAccess)
        {

            UInt32 status = 0;
            IntPtr buffer = IntPtr.Zero, ipHandle = IntPtr.Zero;
            int bufferSize = 0;
            uint pId = 0;
            List<HANDLE_INFO> handlelist = new List<HANDLE_INFO>();
            IntPtr ProcessHandle = IntPtr.Zero, ProcessCopy = IntPtr.Zero;
            ntdll.SYSTEM_HANDLE wHandles = new ntdll.SYSTEM_HANDLE();
            HANDLE_INFO hi;
            long lHandleCount = 0;
            CProcess Process = null;
            do
            {
                status = (UInt32)ntdll.NtQuerySystemInformation(0x10, buffer, bufferSize, ref bufferSize);
                if (status != 0)
                {
                    if (status == 0xc0000004)
                    {
                        if (buffer != null)
                            Kernel32.VirtualFree(buffer, bufferSize, 0x8000);
                        buffer = Kernel32.VirtualAlloc(IntPtr.Zero, bufferSize, 0x1000, 0x40);
                        continue;
                    }
                    break;
                }
                else
                {
                    if (IntPtr.Size == 8)
                    {
                        lHandleCount = Marshal.ReadInt64(buffer);
                        ipHandle = new IntPtr(buffer.ToInt64() + 8);
                    }
                    else
                    {
                        lHandleCount = Marshal.ReadInt32(buffer);
                        ipHandle = new IntPtr(buffer.ToInt32() + 4);
                    }

                    for (long i = 0; i < lHandleCount; i++)
                    {
                        wHandles = new ntdll.SYSTEM_HANDLE();
                        if (IntPtr.Size == 8)
                        {
                            wHandles = (ntdll.SYSTEM_HANDLE)Marshal.PtrToStructure(ipHandle, wHandles.GetType());
                            ipHandle = new IntPtr(ipHandle.ToInt64() + Marshal.SizeOf(wHandles));
                        }
                        else
                        {
                            ipHandle = new IntPtr(ipHandle.ToInt64() + Marshal.SizeOf(wHandles));
                            wHandles = (ntdll.SYSTEM_HANDLE)Marshal.PtrToStructure(ipHandle, wHandles.GetType());
                        }
                        if (wHandles.ObjectTypeNumber == 7 &&
                            wHandles.ProcessID != Kernel32.GetCurrentProcessId() &&
                            (((uint)(wHandles.GrantedAccess) & DesiredAccess)) == DesiredAccess)
                        {
                            if (Options.UseDuplicateHandle == true)
                            {
                                ProcessHandle = Kernel32.OpenProcess(0x0040, false, wHandles.ProcessID);
                                if (Kernel32.DuplicateHandle(ProcessHandle, (IntPtr)(wHandles.Handle), Kernel32.GetCurrentProcess(), ref ProcessCopy, 0x0400, false, 0))
                                {
                                    if (Kernel32.GetProcessId(ProcessCopy) == ProcessId &&
                                        wHandles.ProcessID != ProcessId)
                                    {
                                        hi = new HANDLE_INFO(wHandles.ProcessID,(IntPtr)wHandles.Handle);
                                        handlelist.Add(hi);
                                    }
                                }
                                if (ProcessHandle != IntPtr.Zero)
                                    Kernel32.CloseHandle(ProcessHandle);
                                if (ProcessCopy != IntPtr.Zero)
                                    Kernel32.CloseHandle(ProcessCopy);
                            }
                            else
                            {
                                Process = new CProcess(wHandles.ProcessID, 0x1fffff);
                                if (Process.IsValidProcess() && ServiceGetProcessId(Process, (IntPtr)wHandles.Handle, ref pId))
                                {
                                    if (pId == ProcessId)
                                    {
                                        hi = new HANDLE_INFO(wHandles.ProcessID, (IntPtr)wHandles.Handle);
                                        handlelist.Add(hi);
                                    }
                                }
                                if (Process.IsValidProcess())
                                    Process.Close();
                            }
                        }
                            
                    }
                    break;
                }
            } while (true);
            if (buffer != null)
                Kernel32.VirtualFree(buffer, bufferSize, 0x8000);
            return handlelist;
        }
    }
}

File: HLeaker - c++/HLeaker/Service.cpp
#include "Service.hpp"
#include "Options.hpp"

namespace Service
{
#pragma region Members
	typedef NTSTATUS(NTAPI*_RtlCreateUserThread)(HANDLE Process, PSECURITY_DESCRIPTOR ThreadSecurityDescriptor, BOOLEAN CreateSuspended, ULONG_PTR ZeroBits, SIZE_T MaximumStackSize, SIZE_T CommittedStackSize, PVOID StartAddress, PVOID Parameter, PHANDLE Thread, PVOID ClientId);
	static _RtlCreateUserThread RtlCreateUserThread = reinterpret_cast<_RtlCreateUserThread>(GetProcAddress(GetModuleHandle("ntdll.dll"), "RtlCreateUserThread"));
	typedef NTSTATUS(NTAPI*_NtQuerySystemInformation)(ULONG, PVOID, ULONG, PULONG);
	static _NtQuerySystemInformation NtQuerySystemInformation = reinterpret_cast<_NtQuerySystemInformation>(GetProcAddress(GetModuleHandle("ntdll.dll"), "NtQuerySystemInformation"));
	static void *_ExitThread = GetProcAddress(GetModuleHandle("kernel32.dll"), "ExitThread"), *_GetProcessId = GetProcAddress(GetModuleHandle("kernel32.dll"), "GetProcessId"), *_NtSetInformationObject = GetProcAddress(GetModuleHandleA("ntdll.dll"), "NtSetInformationObject");
#pragma endregion Members

	HANDLE WINAPI ServiceRunProgram(LPCSTR lpFilename, LPCSTR lpArguments, LPCSTR lpDir, LPPROCESS_INFORMATION ProcessInformation, BOOL Inherit, HANDLE hParent)
	{
		HANDLE processToken = NULL, userToken = NULL;
		LPVOID pEnvironment = NULL;
		STARTUPINFOEXA  si = { 0 };
		SIZE_T cbAttributeListSize = 0;
		PPROC_THREAD_ATTRIBUTE_LIST pAttributeList = NULL;
		BOOL Status = TRUE;
		ZeroMemory(&si, sizeof(si));
		si.StartupInfo.cb = sizeof(STARTUPINFOEXA);

		if (!ProcessInformation)
		{
			Status = false;
			goto EXIT;
		}

		if (!OpenProcessToken(GetCurrentProcess(), TOKEN_ALL_ACCESS, &processToken))
		{
			Status = false;
			goto EXIT;
		}
		if (!DuplicateTokenEx(processToken, TOKEN_ALL_ACCESS, NULL, SecurityImpersonation, TokenPrimary, &userToken) ||
			!CreateEnvironmentBlock(&pEnvironment, userToken, TRUE))
		{
			Status = false;
			goto EXIT;
		}
		InitializeProcThreadAttributeList(NULL, 1, 0, &cbAttributeListSize);
		pAttributeList = reinterpret_cast<PPROC_THREAD_ATTRIBUTE_LIST>(HeapAlloc(GetProcessHeap(), 0, cbAttributeListSize));
		if (!pAttributeList)
		{
			Status = false;
			goto EXIT;
		}
		if (!InitializeProcThreadAttributeList(pAttributeList, 1, 0, &cbAttributeListSize))
		{
			Status = false;
			goto EXIT;
		}
		if (!UpdateProcThreadAttribute(pAttributeList, 0, PROC_THREAD_ATTRIBUTE_PARENT_PROCESS, &hParent, sizeof(HANDLE), NULL, NULL))
		{
			Status = false;
			goto EXIT;
		}
		si.lpAttributeList = pAttributeList;
		if (!CreateProcessAsUserA(userToken, lpFilename, const_cast<LPSTR>(lpArguments), NULL, NULL, TRUE, CREATE_UNICODE_ENVIRONMENT | CREATE_NEW_CONSOLE | EXTENDED_STARTUPINFO_PRESENT, pEnvironment, lpDir, reinterpret_cast<LPSTARTUPINFOA>(&si), ProcessInformation))
		{
			Status = false;
			goto EXIT;
		}
	EXIT:
		if (pEnvironment)
			DestroyEnvironmentBlock(pEnvironment);
		CloseHandle(userToken);
		if (ProcessInformation->hThread)
			CloseHandle(ProcessInformation->hThread);
		if (processToken)
			CloseHandle(processToken);
		if (pAttributeList)
		{
			DeleteProcThreadAttributeList(pAttributeList);
			HeapFree(GetProcessHeap(), 0, pAttributeList);
		}
		return Status ? ProcessInformation->hProcess : 0;
	}
	BOOLEAN WINAPI ServiceSetHandleStatus(Process::CProcess* Process, HANDLE hObject, BOOL Protect, BOOL Inherit)
	{
		typedef struct _CLIENT_ID
		{
			HANDLE UniqueProcess;
			HANDLE UniqueThread;

		} CLIENT_ID, *PCLIENT_ID;

		struct HANDLE_IN
		{
			HANDLE hObject;
			BOOL PStatus;
			BOOL IStatus;
			PVOID Function;
		};
#ifdef _WIN64
		static BYTE WThread[] = { 0x48, 0x83, 0xEC, 0x28, 0xF, 0xB6, 0x41, 0x8, 0x4C, 0x8D, 0x44, 0x24, 0x30, 0x41, 0xB9, 0x2, 0x0, 0x0, 0x0, 0x88, 0x44, 0x24, 0x31, 0xF, 0xB6, 0x41, 0xC, 0x4C, 0x8B, 0xD1, 0x48, 0x8B, 0x9, 0x88, 0x44, 0x24, 0x30, 0x41, 0x8D, 0x51, 0x2, 0x41, 0xFF, 0x52, 0x10, 0x33, 0xC9, 0x85, 0xC0, 0xF, 0x94, 0xC1, 0x8B, 0xC1, 0x48, 0x83, 0xC4, 0x28, 0xC3};
#elif _WIN32
		static BYTE WThread[] = { 0x55, 0x8B, 0xEC, 0x8B, 0x4D, 0x8, 0x6A, 0x2, 0xF, 0xB6, 0x41, 0x4, 0x88, 0x45, 0x9, 0xF, 0xB6, 0x41, 0x8, 0x88, 0x45, 0x8, 0x8D, 0x45, 0x8, 0x50, 0x8B, 0x41, 0xC, 0x6A, 0x4, 0xFF, 0x31, 0xFF, 0xD0, 0xF7, 0xD8, 0x1B, 0xC0, 0x40, 0x5D, 0xC2, 0x4, 0x0 };
#endif
		BOOL IsTarget64 = false, Status = false;
		HANDLE_IN Args = { 0,0,0,0 };
		LPVOID lpThread = nullptr, lpArg = nullptr;
		HANDLE hProcess = Process->GetHandle(), hThread = 0;
		int ThreadSize = _countof(WThread);
		if (!RtlCreateUserThread || !_NtSetInformationObject || hProcess == INVALID_HANDLE_VALUE)
			goto EXIT;

		if (!Process->Is64(&IsTarget64))
			goto EXIT;

#ifdef _WIN64
		if (!IsTarget64)
			goto EXIT;
#elif _WIN32
		if (IsTarget64)
			goto EXIT;
#endif
		lpThread = VirtualAllocEx(hProcess, 0, ThreadSize, MEM_COMMIT | MEM_RESERVE, PAGE_EXECUTE_READWRITE);
		if (lpThread == nullptr)
			goto EXIT;

		lpArg = VirtualAllocEx(hProcess, 0, sizeof(HANDLE_IN), MEM_COMMIT | MEM_RESERVE, PAGE_EXECUTE_READWRITE);
		if (lpArg == nullptr)
			goto EXIT;

		Args.Function = _NtSetInformationObject;
		Args.hObject = hObject;
		Args.IStatus = Inherit;
		Args.PStatus = Protect;
		if (!WriteProcessMemory(hProcess, lpThread, reinterpret_cast<LPCVOID>(WThread), ThreadSize, 0) ||
			!WriteProcessMemory(hProcess, lpArg, reinterpret_cast<LPCVOID>(&Args), sizeof(HANDLE_IN), 0))
			goto EXIT;

		if (RtlCreateUserThread(hProcess, 0, 0, 0, 0, 0, reinterpret_cast<PVOID>(lpThread), lpArg, &hThread, 0))
			goto EXIT;

		WaitForSingleObject(hThread, INFINITE);
		Status = true;
	EXIT:
		if (hThread)
			CloseHandle(hThread);
		if (lpThread)
			VirtualFreeEx(hProcess, lpThread, ThreadSize, MEM_RELEASE);
		if (lpArg)
			VirtualFreeEx(hProcess, lpArg, sizeof(HANDLE_IN), MEM_RELEASE);
		return Status;
	}

	BOOLEAN WINAPI ServiceGetProcessId(Process::CProcess* Process, HANDLE hTarget, PDWORD ProcessId)
	{
		struct THREAD_IN
		{
			HANDLE hProcess;
			void* _GetProcessId, *_ExitThread;
		};
#ifdef _WIN64
		static BYTE WThread[] = { 0x40, 0x53, 0x48, 0x83, 0xEC, 0x20, 0x48, 0x8B, 0xD9, 0x48, 0x8B, 0x09, 0xFF, 0x53, 0x08, 0x8B, 0xC8, 0xFF, 0x53, 0x10, 0xB8, 0x01, 0x00, 0x00, 0x00, 0x48, 0x83, 0xC4, 0x20, 0x5B, 0xC3 };
#elif _WIN32
		static BYTE WThread[] = { 0x55, 0x8B, 0xEC, 0x56, 0x8B, 0x75, 0x08, 0xFF, 0x36, 0x8B, 0x46, 0x04, 0xFF, 0xD0, 0x50, 0x8B, 0x46, 0x08, 0xFF, 0xD0, 0x5E, 0x5D, 0xC3 };
#endif
		BOOL IsTarget64 = false, Status = false;
		THREAD_IN Args = { 0,0,0 };
		LPVOID lpThread = nullptr, lpArg = nullptr;
		HANDLE hProcess = Process->GetHandle(), hThread = 0;
		int ThreadSize = _countof(WThread);
		if (!RtlCreateUserThread || !_ExitThread || !_GetProcessId || !ProcessId)
			goto EXIT;

		if (!Process->Is64(&IsTarget64))
			goto EXIT;

#ifdef _WIN64
		if (!IsTarget64)
			goto EXIT;
#elif _WIN32
		if (IsTarget64)
			goto EXIT;
#endif
		lpThread = VirtualAllocEx(hProcess, 0, ThreadSize, MEM_COMMIT | MEM_RESERVE, PAGE_EXECUTE_READWRITE);
		if (lpThread == nullptr)
			goto EXIT;

		lpArg = VirtualAllocEx(hProcess, 0, sizeof(THREAD_IN), MEM_COMMIT | MEM_RESERVE, PAGE_EXECUTE_READWRITE);
		if (lpArg == nullptr)
			goto EXIT;

		Args.hProcess = hTarget;
		Args._ExitThread = _ExitThread;
		Args._GetProcessId = _GetProcessId;
		if (!WriteProcessMemory(hProcess, lpThread, reinterpret_cast<LPCVOID>(WThread), ThreadSize, 0) ||
			!WriteProcessMemory(hProcess, lpArg, reinterpret_cast<LPCVOID>(&Args), sizeof(THREAD_IN), 0))
			goto EXIT;

		if (RtlCreateUserThread(hProcess, 0, 0, 0, 0, 0, reinterpret_cast<PVOID>(lpThread), lpArg, &hThread, 0))
			goto EXIT;
		if (WaitForSingleObject(hThread, OBJECTTIMEOUT) == WAIT_TIMEOUT)
			goto EXIT;
		if (!GetExitCodeThread(hThread, ProcessId))
			goto EXIT;
		Status = true;
	EXIT:
		if (hThread)
			CloseHandle(hThread);
		if (lpThread)
			VirtualFreeEx(hProcess, lpThread, ThreadSize, MEM_RELEASE);
		if (lpArg)
			VirtualFreeEx(hProcess, lpArg, sizeof(THREAD_IN), MEM_RELEASE);
		return Status;
	}

	std::vector<HANDLE_INFO> ServiceEnumHandles(ULONG ProcessId, DWORD dwDesiredAccess)
	{
		typedef struct _SYSTEM_HANDLE
		{
			ULONG ProcessId;
			BYTE ObjectTypeNumber;
			BYTE Flags;
			USHORT Handle;
			PVOID Object;
			ACCESS_MASK GrantedAccess;
		} SYSTEM_HANDLE, *PSYSTEM_HANDLE;

		typedef struct _SYSTEM_HANDLE_INFORMATION
		{
			ULONG HandleCount;
			SYSTEM_HANDLE Handles[1];
		} SYSTEM_HANDLE_INFORMATION, *PSYSTEM_HANDLE_INFORMATION;

		PSYSTEM_HANDLE_INFORMATION handleInfo = 0;
		NTSTATUS status = -1;
		PVOID buffer = 0;
		ULONG bufferSize = 0, pId = 0;
		std::vector<HANDLE_INFO> handlelist;
		HANDLE ProcessHandle = 0, ProcessCopy = 0;
		HANDLE_INFO hi = { 0,0 };
		Process::CProcess* Process = 0;

		if (!NtQuerySystemInformation)
			goto EXIT;

		do {
			status = NtQuerySystemInformation(0x10, buffer, bufferSize, &bufferSize);
			if (status) {
				if (status == 0xc0000004) {
					if (buffer != NULL)
						VirtualFree(buffer, bufferSize, MEM_DECOMMIT);
					buffer = VirtualAlloc(0, bufferSize, MEM_COMMIT, PAGE_EXECUTE_READWRITE);
					continue;
				}
				break;
			}
			else {
				handleInfo = reinterpret_cast<PSYSTEM_HANDLE_INFORMATION>(buffer);
				for (int i = 0; i < handleInfo->HandleCount; i++) {
					auto handle = &handleInfo->Handles[i];
					if (handle->ObjectTypeNumber == 7 && (handle->GrantedAccess & dwDesiredAccess) == dwDesiredAccess)
					{
#if (USE_DUPLICATE_HANDLE == 1)
						ProcessHandle = OpenProcess(PROCESS_DUP_HANDLE, false, handle->ProcessId);
						if (DuplicateHandle(ProcessHandle, reinterpret_cast<HANDLE>(handle->Handle), GetCurrentProcess(), &ProcessCopy, PROCESS_QUERY_INFORMATION, 0, 0))
						{
							if (GetProcessId(ProcessCopy) == ProcessId)
								handlelist.push_back(HANDLE_INFO(handle->ProcessId, reinterpret_cast<HANDLE>(handle->Handle)));
							
						}
						if (ProcessHandle)
							CloseHandle(ProcessHandle);
						if (ProcessCopy)
							CloseHandle(ProcessCopy);
#else
						Process = new Process::CProcess(handle->ProcessId, PROCESS_ALL_ACCESS);
						if (Process->IsValidProcess() && ServiceGetProcessId(Process, reinterpret_cast<HANDLE>(handle->Handle), &pId))
						{
							if (pId == ProcessId)
								handlelist.push_back(HANDLE_INFO(handle->ProcessId, reinterpret_cast<HANDLE>(handle->Handle)));
							
						}
						if (Process->IsValidProcess())
							Process->Close();
						delete Process;
#endif

					}
				}
				break;
			}
		} 
		while (true);
	EXIT:
		if (buffer != NULL)
			VirtualFree(buffer, bufferSize, MEM_DECOMMIT);
		return handlelist;
	}
}
		
